In [1]:
# Library to suppress warnings or deprecation notes 
import warnings
warnings.filterwarnings('ignore')

# Libraries to help with reading and manipulating data
import numpy as np
import pandas as pd

# Libraries to help with data visualization
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

# Libraries to split data, impute missing values 
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer

# Libraries to import decision tree classifier and different ensemble classifiers
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier, GradientBoostingClassifier
from xgboost import XGBClassifier
from sklearn.ensemble import StackingClassifier
from sklearn.tree import DecisionTreeClassifier

# Libtune to tune model, get different metric scores
from sklearn import metrics
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score, precision_score, recall_score
from sklearn.metrics import f1_score, roc_auc_score
from sklearn.model_selection import GridSearchCV

In [2]:
df = pd.read_excel('Tourism.xlsx', sheet_name='Tourism')

In [3]:
df.shape

(4888, 20)

In [4]:
dic = pd.read_excel("Tourism.xlsx")

In [5]:
dic.shape

(21, 4)

In [6]:
dic.head(5)

Unnamed: 0 Unnamed: 1     Unnamed: 2  \
0         NaN       Data       Variable   
1         NaN    Tourism     CustomerID   
2         NaN    Tourism      ProdTaken   
3         NaN    Tourism            Age   
4         NaN    Tourism  TypeofContact   

                                          Unnamed: 3  
0                                        Discerption  
1                                 Unique customer ID  
2  Whether the customer has purchased a package o...  
3                                    Age of customer  
4  How customer was contacted (Company Invited or...

In [7]:
dic = dic.drop(['Unnamed: 0','Unnamed: 1'], axis=1)
dic = dic.drop(0, axis=0)
dic.columns = ['Variable','Description']

In [8]:
dic.sample()

Variable                                        Description
9  NumberOfPersonVisiting  Total number of persons planning to take the t...

In [9]:
df.head()

CustomerID  ProdTaken   Age    TypeofContact  CityTier  DurationOfPitch  \
0      200000          1  41.0     Self Enquiry         3              6.0   
1      200001          0  49.0  Company Invited         1             14.0   
2      200002          1  37.0     Self Enquiry         1              8.0   
3      200003          0  33.0  Company Invited         1              9.0   
4      200004          0   NaN     Self Enquiry         1              8.0   

       Occupation  Gender  NumberOfPersonVisiting  NumberOfFollowups  \
0        Salaried  Female                       3                3.0   
1        Salaried    Male                       3                4.0   
2     Free Lancer    Male                       3                4.0   
3        Salaried  Female                       2                3.0   
4  Small Business    Male                       2                3.0   

  ProductPitched  PreferredPropertyStar MaritalStatus  NumberOfTrips  \
0         Deluxe                    3.0        Single            1.0   
1         Deluxe                    4.0      Divorced            2.0   
2          Basic                    3.0        Single            7.0   
3          Basic                    3.0      Divorced            2.0   
4          Basic                    4.0      Divorced            1.0   

   Passport  PitchSatisfactionScore  OwnCar  NumberOfChildrenVisiting  \
0         1                       2       1                       0.0   
1         0                       3       1                       2.0   
2         1                       3       0                       0.0   
3         1                       5       1                       1.0   
4         0                       5       1                       0.0   

  Designation  MonthlyIncome  
0     Manager        20993.0  
1     Manager        20130.0  
2   Executive        17090.0  
3   Executive        17909.0  
4   Executive        18468.0

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4888 entries, 0 to 4887
Data columns (total 20 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CustomerID                4888 non-null   int64  
 1   ProdTaken                 4888 non-null   int64  
 2   Age                       4662 non-null   float64
 3   TypeofContact             4863 non-null   object 
 4   CityTier                  4888 non-null   int64  
 5   DurationOfPitch           4637 non-null   float64
 6   Occupation                4888 non-null   object 
 7   Gender                    4888 non-null   object 
 8   NumberOfPersonVisiting    4888 non-null   int64  
 9   NumberOfFollowups         4843 non-null   float64
 10  ProductPitched            4888 non-null   object 
 11  PreferredPropertyStar     4862 non-null   float64
 12  MaritalStatus             4888 non-null   object 
 13  NumberOfTrips             4748 non-null   float64
 14  Passport

In [11]:
df = df.drop(columns='CustomerID')

In [12]:
df.describe().T

count          mean          std     min      25%  \
ProdTaken                 4888.0      0.188216     0.390925     0.0      0.0   
Age                       4662.0     37.622265     9.316387    18.0     31.0   
CityTier                  4888.0      1.654255     0.916583     1.0      1.0   
DurationOfPitch           4637.0     15.490835     8.519643     5.0      9.0   
NumberOfPersonVisiting    4888.0      2.905074     0.724891     1.0      2.0   
NumberOfFollowups         4843.0      3.708445     1.002509     1.0      3.0   
PreferredPropertyStar     4862.0      3.581037     0.798009     3.0      3.0   
NumberOfTrips             4748.0      3.236521     1.849019     1.0      2.0   
Passport                  4888.0      0.290917     0.454232     0.0      0.0   
PitchSatisfactionScore    4888.0      3.078151     1.365792     1.0      2.0   
OwnCar                    4888.0      0.620295     0.485363     0.0      0.0   
NumberOfChildrenVisiting  4822.0      1.187267     0.857861     0.0      1.0   
MonthlyIncome             4655.0  23619.853491  5380.698361  1000.0  20346.0   

                              50%      75%      max  
ProdTaken                     0.0      0.0      1.0  
Age                          36.0     44.0     61.0  
CityTier                      1.0      3.0      3.0  
DurationOfPitch              13.0     20.0    127.0  
NumberOfPersonVisiting        3.0      3.0      5.0  
NumberOfFollowups             4.0      4.0      6.0  
PreferredPropertyStar         3.0      4.0      5.0  
NumberOfTrips                 3.0      4.0     22.0  
Passport                      0.0      1.0      1.0  
PitchSatisfactionScore        3.0      4.0      5.0  
OwnCar                        1.0      1.0      1.0  
NumberOfChildrenVisiting      1.0      2.0      3.0  
MonthlyIncome             22347.0  25571.0  98678.0

In [49]:

# Generating a data frame of missing values by count and %
missing = (df.isnull().sum()/df.isnull().count()*100).sort_values(ascending=False)
missing = pd.DataFrame(missing, columns=['%'])
missing['count'] = df.isnull().sum()


In [54]:
print('Missing Values in the data frame')
missing.head()

Missing Values in the data frame


%  count
DurationOfPitch           5.135025    251
MonthlyIncome             4.766776    233
Age                       4.623568    226
NumberOfTrips             2.864157    140
NumberOfChildrenVisiting  1.350245     66

In [55]:
print('The count of unique values in the data frame')
df.nunique().sort_values(ascending=False)

The count of unique values in the data frame


MonthlyIncome               2475
Age                           44
DurationOfPitch               34
NumberOfTrips                 12
NumberOfFollowups              6
Designation                    5
PitchSatisfactionScore         5
ProductPitched                 5
NumberOfPersonVisiting         5
MaritalStatus                  4
Occupation                     4
NumberOfChildrenVisiting       4
Gender                         3
PreferredPropertyStar          3
CityTier                       3
Passport                       2
OwnCar                         2
TypeofContact                  2
ProdTaken                      2
dtype: int64

In [58]:
dic.head(21)

Variable  \
1                 CustomerID   
2                  ProdTaken   
3                        Age   
4              TypeofContact   
5                   CityTier   
6            DurationOfPitch   
7                 Occupation   
8                     Gender   
9     NumberOfPersonVisiting   
10         NumberOfFollowups   
11            ProductPitched   
12     PreferredPropertyStar   
13             MaritalStatus   
14             NumberOfTrips   
15                  Passport   
16    PitchSatisfactionScore   
17                    OwnCar   
18  NumberOfChildrenVisiting   
19               Designation   
20             MonthlyIncome   

                                          Description  
1                                  Unique customer ID  
2   Whether the customer has purchased a package o...  
3                                     Age of customer  
4   How customer was contacted (Company Invited or...  
5   City tier depends on the development of a city...  
6   Duration of the pitch by a salesperson to the ...  
7                              Occupation of customer  
8                                  Gender of customer  
9   Total number of persons planning to take the t...  
10  Total number of follow-ups has been done by sa...  
11                 Product pitched by the salesperson  
12        Preferred hotel property rating by customer  
13                         Marital status of customer  
14      Average number of trips in a year by customer  
15  The customer has a passport or not (0: No, 1: ...  
16                     Sales pitch satisfaction score  
17  Whether the customers own a car or not (0: No,...  
18  Total number of children with age less than 5 ...  
19  Designation of the customer in the current org...  
20               Gross monthly income of the customer

In [73]:
# Printing a list of unique values, values shown in text file. Some categories have a large number of values so dropping those
vc = df.columns.drop(['MonthlyIncome','Age','DurationOfPitch','NumberOfTrips'])
for i in vc:
    print(i,'Has the following Unique Values')
    print(df[i].value_counts().sort_values(ascending=False))
    print('-'*50)

ProdTaken Has the following Unique Values
0    3968
1     920
Name: ProdTaken, dtype: int64
--------------------------------------------------
TypeofContact Has the following Unique Values
Self Enquiry       3444
Company Invited    1419
Name: TypeofContact, dtype: int64
--------------------------------------------------
CityTier Has the following Unique Values
1    3190
3    1500
2     198
Name: CityTier, dtype: int64
--------------------------------------------------
Occupation Has the following Unique Values
Salaried          2368
Small Business    2084
Large Business     434
Free Lancer          2
Name: Occupation, dtype: int64
--------------------------------------------------
Gender Has the following Unique Values
Male       2916
Female     1817
Fe Male     155
Name: Gender, dtype: int64
--------------------------------------------------
NumberOfPersonVisiting Has the following Unique Values
3    2402
2    1418
4    1026
1      39
5       3
Name: NumberOfPersonVisiting, dtype: int

### Data Intial observations
 * CustomerID is redundant with the index
 <br/><br/>
 * The following columns have missing values we will need to fix
    - DurationOfPitch - replace with mean
    - MonthlyIncome - replace with mean
    - Age - replace with mean
    - NumberOfTrips - replace with mode
    - NumberOfChildernVisting - replace with mode
    <br/><br/>
 * The following categorical columns can be dummies
    - TypeofContact
    - Occupation
    - Gender
    - ProductPitched
    - MartialStatus
    - Designation
    <br/><br/>
 * Customer Obeseravation:
    - 50% of customers are between 31 & 44
    - Most pitches are between 9 & 20 minutes, with the average being 15 min
    - 75% of trips have 1 - 3 people, only 25% of trips have 1 person
    - 25% of customers buy on the 1st Follow up, Most customers need 3-4 follow ups
    - No customers will book a property with less than 3 stars
    - 75% of customers take 1 - 4 trips per year
    - 75% of customers do not have passports making most trips local
    - Pitch Satification Score will be interesting during BiVariate Analysis
    - 50% of customers make between 20,300 - 25,600 per month
    <br/><br/>
 * Possible Outliers:
    - DurationOfPitch - max of 127
    - NumberOfTrips - 22 on the High End
    - MonthlyIncome - 98,678 might not be an outlier just a high income
    <br/><br/>
 * General Objservations:
    - roughly 29% of customers were invited the rest self inquired.
    - With that many customers self selecting advertising to select audiences might be a much stronger strategy instead of targeted reach outs.
    - The fast majority of visits are too tier 1 cities just under 2/3rd, with roughly 1/3rd traveling to tier 3 cities. There are almost no trips to tier 2 cities. The company should depri advertising to tier 2 cities
    - Married customers make up almost as much of the population the buys a package as all other categories. Advertising should prioritize married customers
    - The vast majority of packages are purchased by Executives and Managers. Advertising should prioritize these customers
    - The Gender Column has a typo creating an additional "Fe Male" value. 

In [76]:
# Fixing the gender columns
df['Gender'] = df['Gender'].replace(['Fe Male'],'Female')
df['Gender'].value_counts()

Male      2916
Female    1972
Name: Gender, dtype: int64